src/Workspaces/Core/Portable/SymbolKey/SymbolKey.NamedTypeSymbolKey.cs

```diff
@@ -34,30 +31,43 @@ public static SymbolKeyResolution Resolve(SymbolKeyReader reader)
                 var metadataName = reader.ReadString();
                 var containingSymbolResolution = reader.ReadSymbolKey();
                 var arity = reader.ReadInteger();
-                var typeKind = (TypeKind)reader.ReadInteger();
                 var isUnboundGenericType = reader.ReadBoolean();
-                var typeArgumentsOpt = reader.ReadSymbolKeyArray();
+                using var typeArguments = reader.ReadSymbolKeyArray<ITypeSymbol>();
 
-                var types = GetAllSymbols<INamespaceOrTypeSymbol>(containingSymbolResolution).SelectMany(
-                    s => Resolve(reader, s, metadataName, arity, typeKind, isUnboundGenericType, typeArgumentsOpt));
-                return CreateSymbolInfo(types);
+                if (typeArguments.IsDefault)
+                {
+                    return default;
+                }
+
+                var typeArgumentArray = typeArguments.Count == 0
+                    ? Array.Empty<ITypeSymbol>()
+                    : typeArguments.Builder.ToArray();
+                using var result = PooledArrayBuilder<INamedTypeSymbol>.GetInstance();
+                foreach (var nsOrType in containingSymbolResolution.OfType<INamespaceOrTypeSymbol>())
+                {
+                    Resolve(
+                        result, nsOrType, metadataName, arity,
+                        isUnboundGenericType, typeArgumentArray);
+                }
+
+                return CreateSymbolInfo(result);
             }
-            private static IEnumerable<INamedTypeSymbol> Resolve(
-                SymbolKeyReader reader,
+
+            private static void Resolve(
+                PooledArrayBuilder<INamedTypeSymbol> result,
                 INamespaceOrTypeSymbol container,
                 string metadataName,
                 int arity,
-                TypeKind typeKind,
                 bool isUnboundGenericType,
-                ImmutableArray<SymbolKeyResolution> typeArguments)
+                ITypeSymbol[] typeArguments)
             {
-                var types = container.GetTypeMembers(GetName(metadataName), arity);
-                var result = InstantiateTypes(
-                    reader.Compilation, reader.IgnoreAssemblyKey, types, arity, typeArguments);
+                foreach (var type in container.GetTypeMembers(GetName(metadataName), arity))
+                {
+                    var currentType = typeArguments.Length > 0 ? type.Construct(typeArguments) : type;
+                    currentType = isUnboundGenericType ? currentType.ConstructUnboundGenericType() : currentType;
 
-                return isUnboundGenericType
-                    ? result.Select(t => t.ConstructUnboundGenericType())
-                    : result;
+                    result.AddIfNotNull(currentType);
```

Comment:
>currentType [](start = 40, length = 11)

can `currentType` be null?

Reply:
it cannot be.  however, i didn't want to expose .Add and .AddIfNotNull.  Because then the caller woudl have to pick, and they might get it wrong.  so i *only* exposed AddIfNotNull to make it clear that you're *always* getting those semantics here.  The array will *never* have null in it.